watertap/unit_models/tests/unit_test_harness.py

```diff
@@ -0,0 +1,186 @@
+            )
+        if number_total_constraints(blk) != stats["number_total_constraints"]:
+            raise UnitValueError(
+                "The number of constraints were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_total_constraints(blk),
+                    num_test=stats["number_total_constraints"],
+                )
+            )
+        if number_unused_variables(blk) != stats["number_unused_variables"]:
+            raise UnitValueError(
+                "The number of unused variables were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_unused_variables(blk),
+                    num_test=stats["number_unused_variables"],
+                )
+            )
+
+    @pytest.mark.unit
+    def test_units_consistent(self, frame_unit):
+        assert_units_consistent(frame_unit)
+
+    @pytest.mark.unit
+    def test_dof(self, frame_unit):
+        if degrees_of_freedom(frame_unit) != 0:
+            raise UnitAttributeError(
+                "The unit has {dof} degrees of freedom when 0 is required."
+                "".format(dof=degrees_of_freedom(frame_unit))
+            )
+
+    @pytest.mark.component
+    def test_initialization(self, frame_unit):
+        blk = frame_unit
+
+        # initialize
+        try:
+            blk.initialize(solver=blk._test_objs.solver, optarg=blk._test_objs.optarg)
+        except InitializationError:
+            raise InitializationError(
+                "The unit has failed to initialize successfully. Please check the output logs for more information."
+            )
+
+    @pytest.mark.component
+    def test_unit_solutions(self, frame_unit):
+        self.configure_class()
+        blk = frame_unit
+        solutions = blk._test_objs.unit_solutions
+
+        # solve unit
+        if blk._test_objs.solver is None:
+            opt = get_solver()
+        else:
+            opt = get_solver(
+                solver=blk._test_objs.solver, options=blk._test_objs.optarg
+            )
+        results = opt.solve(blk)
+
+        # check solve
+        try:
+            assert len(list(iscale.badly_scaled_var_generator(blk, zero=1e-8))) == 0
+        except AssertionError:
+            badly_scaled_var_list = iscale.badly_scaled_var_generator(blk, zero=1e-8)
+            for x in badly_scaled_var_list:
+                print(
+                    f"{x[0].name}\t{x[0].value}\tsf: {iscale.get_scaling_factor(x[0])}"
+                )
+            raise AssertionError(
+                "Check the output logs for variables that need to be re-scaled"
+            )
+        assert_optimal_termination(results)
+
+        # check results
+
+        for key, val in solutions.items():
+            try:
+                var_name = str(key)
+                truncated_var_name = var_name[8:]
+                pyo_obj = blk.find_component(truncated_var_name)
+                pyo_obj_value = pyo_obj()
```

Comment:
Not sure how to get around hard-coding `truncated_var_name` because `find_component` won't work if the variable starts with "m.fs.unit"

Reply:
Never mind, the key is actually the variable. So you don't need these lines at all, just get its value.